dask/array/gufunc.py

```diff
@@ -0,0 +1,434 @@
+        new core dimensions (not found on inputs) appear on outputs.
+    vectorize: bool, keyword only
+        If set to ``True``, ``np.vectorize`` is applied to ``func`` for
+        convenience. Defaults to ``False``.
+    allow_rechunk: Optional, bool, keyword only
+        Allows rechunking, otherwise chunk sizes need to match and core
+        dimensions are to consist only of one chunk.
+        Warning: enabling this can increase memory usage significantly.
+        Defaults to ``False``.
+    **kwargs : dict
+        Extra keyword arguments to pass to `func`
+
+    Returns
+    -------
+    Single dask.array.Array or tuple of dask.array.Array
+
+    Examples
+    --------
+    >>> import dask.array as da
+    >>> import numpy as np
+    >>> def stats(x):
+    ...     return np.mean(x, axis=-1), np.std(x, axis=-1)
+    >>> a = da.random.normal(size=(10,20,30), chunks=(5, 10, 30))
+    >>> mean, std = da.apply_gufunc(stats, "(i)->(),()", a, output_dtypes=2*(a.dtype,))
+    >>> mean.compute().shape
+    (10, 20)
+
+
+    >>> def outer_product(x, y):
+    ...     return np.einsum("i,j->ij", x, y)
+    >>> a = da.random.normal(size=(   20,30), chunks=(10, 30))
+    >>> b = da.random.normal(size=(10, 1,40), chunks=(5, 1, 40))
+    >>> c = da.apply_gufunc(outer_product, "(i),(j)->(i,j)", a, b, output_dtypes=a.dtype, vectorize=True)
+    >>> c.compute().shape
+    (10, 20, 30, 40)
+
+    References
+    ----------
+    .. [1] http://docs.scipy.org/doc/numpy/reference/ufuncs.html
+    .. [2] http://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html
+    """
+    output_dtypes = kwargs.pop("output_dtypes", None)
+    output_sizes = kwargs.pop("output_sizes", None)
+    vectorize = kwargs.pop("vectorize", None)
+    allow_rechunk = kwargs.pop("allow_rechunk", False)
+
+    # Input processing:
+    ## Signature
+    if not isinstance(signature, str):
+        raise ValueError('`signature` has to be of type string')
+    core_input_dimss, core_output_dimss = _parse_gufunc_signature(signature)
+
+    ## Determine nout: nout = None for functions of one direct return; nout = int for return tuples
+    nout = None if not isinstance(core_output_dimss, list) else len(core_output_dimss)
+
+    ## Assert output_dtypes
+    if output_dtypes is None:
+        raise ValueError("Must specify `output_dtypes` of output array(s)")
+    elif isinstance(output_dtypes, str):
+        otypes = list(output_dtypes)
+        output_dtypes = otypes[0] if nout is None else otypes
+    elif isinstance(output_dtypes, (tuple, list)):
+        if nout is None:
+            raise ValueError("Must specify single dtype for `output_dtypes` for function with one output")
+        otypes = output_dtypes
+    else:
+        if nout is not None:
+            raise ValueError("Must specify tuple of dtypes for `output_dtypes` for function with multiple outputs")
+        otypes = [output_dtypes]
```

Comment:
I don't quite understand why all this logic is necessary. Are you trying to be more flexible in allowing a single output_dtype to be provided if there is only one output?